pwiz_tools/Skyline/Model/SmallMoleculeTransitionListReader.cs

```diff
@@ -577,47 +577,102 @@ private int INDEX_DECLUSTERING_POTENTIAL
+            var mass = mz.IsMonoIsotopic() ? monoMass : averageMass;
+
             // Does given charge, if any, agree with mass and mz?
-            if (adduct.IsEmpty && charge.HasValue)
+            var adductInferred = adduct;
+            if (adduct.IsEmpty && (charge??0) != 0)
             {
-                adduct = Adduct.NonProteomicProtonatedFromCharge(charge.Value);
+                adductInferred = Adduct.NonProteomicProtonatedFromCharge(charge.Value);
             }
-            mzCalc = adduct.AdductCharge != 0 ? adduct.MzFromNeutralMass(mass) : (double?) null;
+            mzCalc = adductInferred.AdductCharge != 0 ? adductInferred.MzFromNeutralMass(mass) : (double?) null;
             if (mzCalc.HasValue && tolerance >= (Math.Abs(mzCalc.Value - mz)))
             {
+                adduct = adductInferred;
                 return charge;
             }
-            int nearestCharge;
-            var calculatedCharge = TransitionCalc.CalcCharge(mass, mz, tolerance, true,
-                TransitionGroup.MIN_PRECURSOR_CHARGE,
-                TransitionGroup.MAX_PRECURSOR_CHARGE, new int[0],
-                TransitionCalc.MassShiftType.none, out massShift, out nearestCharge);
-            if (calculatedCharge.IsEmpty)
+
+            // See if this can be explained by (de)protonation within acceptable charge range
+            var minCharge = (isPositive ?? false)
+                ? TransitionGroup.MIN_PRECURSOR_CHARGE
+                : -TransitionGroup.MAX_PRECURSOR_CHARGE;
+            var maxCharge = (isPositive ?? true)
+                ? TransitionGroup.MAX_PRECURSOR_CHARGE
+                : -TransitionGroup.MIN_PRECURSOR_CHARGE;
+            adductInferred = TransitionCalc.CalcCharge(mass, mz, tolerance, true,
+                minCharge,
+                maxCharge, new int[0],
+                TransitionCalc.MassShiftType.none, out _, out _);
+
+            if (isPrecursor && adductInferred.IsEmpty)
+            {
+                // See if this can be explained by the more common adduct types, possibly with water loss
+                var matches = new Dictionary<double, Adduct>();
+                foreach (var text in Adduct.DEFACTO_STANDARD_ADDUCTS)
+                {
+                    adductInferred = Adduct.FromString(text, Adduct.ADDUCT_TYPE.non_proteomic, null);
+                    if (minCharge <= adductInferred.AdductCharge && adductInferred.AdductCharge <= maxCharge)
+                    {
+                        var err = Math.Abs(adductInferred.MzFromNeutralMass(mass) - mz);
+                        if (err <= tolerance)
+                        {
+                            matches.Add(err, adductInferred);
+                        }
+                        else
+                        {
+                            // Try water loss
+                            var parts = text.Split('+', '-'); // Only for simple adducts like M+H, M+Na etc
+                            if (parts.Length == 2)

```

Comment:
Do we anticipate ever needing to support more than water loss and double water loss? If so it could be helpful to rewrite this in a more generalized form.

Reply:
Possibly? It seems like most other possibilities are already covered by that list of "most common" adducts put out by the Fiehn lab.